devlib/module/sched.py

```diff
@@ -197,24 +197,108 @@ def check_version(cls, target, logger):
                 "but target is running v{}".format(ref_parts, parts)
             )
 
-
     def __str__(self):
         return self.name
 
 
+class _SchedDomainFlagMeta(type):
+    """
+    Metaclass of :class:`SchedDomainFlag`.
+
+    Provides some level of emulation of :class:`enum.Enum` behavior for
+    backward compatibility.
+    """
+    @property
+    def _flags(self):
+        return [
+            attr
+            for name, attr in self.__dict__.items()
+            if name.startswith('SD_')
+        ]
+
+    def __getitem__(self, i):
+        return self._flags[i]
+
+    def __len__(self, i):
+        return len(self._flags)
+
+    # These would be provided by collections.abc.Sequence, but using it on a
+    # metaclass seems to have issues around __init_subclass__
+    def __iter__(self, i):
```

Comment:
Testing this on a v4.14 target yields:
```
sd_info = target.sched.get_sd_info()
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 484, in get_sd_info
    return SchedProcFSData(self.target)
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 332, in __init__
    super(SchedProcFSData, self).__init__(procfs)
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 139, in __init__
    self._dyn_attrs[dest][i] = self._build_node(key, nodes[key])
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 115, in _build_node
    return SchedProcFSNode._build_directory(node_name, node_data)
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 98, in _build_directory
    return SchedProcFSNode(node_data)
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 139, in __init__
    self._dyn_attrs[dest][i] = self._build_node(key, nodes[key])
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 115, in _build_node
    return SchedProcFSNode._build_directory(node_name, node_data)
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 96, in _build_directory
    return SchedDomain(node_data)
  File "/data/work/lisa/external/devlib/devlib/module/sched.py", line 297, in __init__
    for flag in SchedDomainFlag
TypeError: __iter__() missing 1 required positional argument: 'i'
```

AFAICT none of the methods here except `__getitem__()` should have this `i` parameter.

Other than this, I believe this is the right approach, and it works fine on a bleeding edge kernel.

Reply:
ah right, copy paste issue, strange that I haven't hit it in my tests, thanks